neuralpp/experiments/experimental_inference/belief_propagation.py

```diff
@@ -0,0 +1,38 @@
+from typing import List, Optional
+
+from functools import reduce
+from neuralpp.inference.graphical_model.representation.factor.product_factor import (
+    Factor
+)
+from neuralpp.inference.graphical_model.variable.discrete_variable import DiscreteVariable
+from neuralpp.inference.graphical_model.variable.variable import Variable
+
+
+class BeliefPropagation:
+
+    @staticmethod
+    def mult_factors(factors: List[Factor]):
+        return reduce(lambda x, y: x * y, factors)
+
+    @staticmethod
+    def add_factors(factors: List[Factor]):
+        return reduce(lambda x, y: x.add_with_non_identity(y), factors)
+
+    @staticmethod
+    def run(query: Variable, factors: List[Factor]):
+        return BeliefPropagation.msg_from_variable(query, None, factors)
+
+    @staticmethod
+    def msg_from_factor(factor: Factor, out_var: Variable, factors: List[Factor]):
+        neighbors = [v for v in factor.variables if v != out_var]
+        if not neighbors:
+            return factor
+        assignments = DiscreteVariable.assignments_product_dicts(neighbors)
+        msg = BeliefPropagation.mult_factors([BeliefPropagation.msg_from_variable(v, factor, factors) for v in neighbors])
+        return BeliefPropagation.add_factors([msg.condition(a) * factor.condition(a) for a in assignments])
```

Comment:
You don't need to iterate over assignments here. In fact, this may be wasteful because sometimes we can compute  `sum_a f(a)`  without iterating over all `a`'s (for example if `f(a)` is `if a = 1 then 0.1 else 0.2`.

To compute a summation `phi_2(w) = sum_{x,y,z} phi_1(x,y,z,w)` over a factor in a way that leverages the factor's internal representation (without having to know what that internal representation is), you can use `phi_1.sum_out([x,y,z])` instead. This will return factor `phi_2` defined on variable `w`.

In your scenario here you will invoke `sum_out` on the `ProductFactor` you created above.